### ISSUE TITLE
query online is slow, is there a better way to do it?

### ISSUE THREAD

**Issue by funwun** [erlio/vernemq#433]
### Environment

- VerneMQ Version: 1.1.0
- OS: ubuntu 16.04
- Erlang/OTP version (if building from source):

### Expected behavior

I have use this two ways to get the status of session

```
sudo vmq-admin session show --client_id=test --is_online=true
```

```
http://xxxxxxxx@localhost:8888/api/v1/session/show?--client_id=test
```

curl elapsed time
```
time_namelookup:  0.000
time_connect:  0.002
time_appconnect:  0.000
time_pretransfer:  0.006
time_redirect:  0.000
 time_starttransfer:  0.812
```

I know the two ways is the same process.
I know vmq_ql traverses all sessions (and queues) and of course takes longer to execute.

But when I have 10k clients connect to the broker, the query is too slow, is there a better way to do this?

I don't know vernemq is how to store the session status list, can you explain this why is to slow?

**Comment by dcy** [comment]
If you develope your own plugin, you can:
```
is_online(ClientId) ->
    is_online(get_nodes(), {[], ClientId}).

is_online([], _SubscriberId) ->
    false;
is_online([Node | Nodes], SubscriberId) when Node == node() ->
    case is_online_(SubscriberId) of
        true -> true;
        false -> is_online(Nodes, SubscriberId)
    end;
is_online([Node | Nodes], SubscriberId) ->
    case rpc:call(Node, ?MODULE, is_online_, [SubscriberId]) of
        true -> true;
        false -> is_online(Nodes, SubscriberId)
    end.

is_online_(SubscriberId) ->
    case vmq_queue_sup_sup:get_queue_pid(SubscriberId) of
        not_found ->
            false;
        QPid ->
            case vmq_queue:status(QPid) of
                {online, _Mode, _TotalStoredMsgs, _Sessions, _IsPlugin} -> true;
                _ -> false
            end
    end.

get_nodes() ->
    LocalNode = node(),
    [LocalNode | (vmq_cluster:nodes() -- [LocalNode])].
```

**Comment by dergraf** [comment]
@funwun I don't understand the metric you have provided. How long does it take? several milliseconds or multiple seconds?

**Comment by funwun** [comment]
800ms~1200ms

**Comment by falood** [comment]
the same issue for me.
`vmq-admin session show` works well but when I add a filter like `--user=my_own_username`, it will be very slow, get `RPC to 'VerneMQ@127.0.0.1' failed: timeout`

**Comment by larshesel** [comment]
What is the load when you experience this? Are there some clients publishing or receiving many msgs/s?

**Comment by falood** [comment]
@larshesel very low load, only few messages publishing and receiving.

**Comment by larshesel** [comment]
Could you describe your setup a bit more when you see this? Is it always happening or only sometimes? If it's only sometimes are there anything unusual when it happens compared to when it doesn't happen? Any context would be helpful.

**Comment by falood** [comment]
1. every time I restarted vernemq server, everything works well
2. when I restart vernemq server wieh this issue, I got such error log 
```
==> log/console.log <==
2017-07-24 21:02:08.259 [error] <0.113.0> Supervisor vmq_ql_query_sup had child vmq_ql_query started with {vmq_ql_query,start_link,undefined} at <0.12482.18> exit with reason {error,badarg} in context child_terminated

==> log/erlang.log.1 <==

==> log/error.log <==
2017-07-24 21:02:08.259 [error] <0.113.0> Supervisor vmq_ql_query_sup had child vmq_ql_query started with {vmq_ql_query,start_link,undefined} at <0.12442.18> exit with reason {error,badarg} in context child_terminated

==> log/console.log <==
2017-07-24 21:02:08.259 [error] <0.113.0> Supervisor vmq_ql_query_sup had child vmq_ql_query started with {vmq_ql_query,start_link,undefined} at <0.12442.18> exit with reason {error,badarg} in context child_terminated

==> log/error.log <==
2017-07-24 21:02:08.259 [error] <0.113.0> Supervisor vmq_ql_query_sup had child vmq_ql_query started with {vmq_ql_query,start_link,undefined} at <0.12464.18> exit with reason {error,badarg} in context child_terminated

==> log/console.log <==
2017-07-24 21:02:08.259 [error] <0.113.0> Supervisor vmq_ql_query_sup had child vmq_ql_query started with {vmq_ql_query,start_link,undefined} at <0.12464.18> exit with reason {error,badarg} in context child_terminated

==> log/error.log <==
2017-07-24 21:02:08.259 [error] <0.12431.18> gen_server <0.12431.18> terminated with reason: {error,badarg} in vmq_ql_query:select/5 line 187

==> log/console.log <==
2017-07-24 21:02:08.259 [error] <0.12431.18> gen_server <0.12431.18> terminated with reason: {error,badarg} in vmq_ql_query:select/5 line 187

==> log/error.log <==
2017-07-24 21:02:08.259 [error] <0.12431.18> CRASH REPORT Process <0.12431.18> with 0 neighbours exited with reason: {error,badarg} in gen_server:terminate/7 line 812

==> log/console.log <==
2017-07-24 21:02:08.259 [error] <0.12431.18> CRASH REPORT Process <0.12431.18> with 0 neighbours exited with reason: {error,badarg} in gen_server:terminate/7 line 812

==> log/error.log <==
2017-07-24 21:02:08.259 [error] <0.113.0> Supervisor vmq_ql_query_sup had child vmq_ql_query started with {vmq_ql_query,start_link,undefined} at <0.12431.18> exit with reason {error,badarg} in context child_terminated

==> log/console.log <==
2017-07-24 21:02:08.259 [error] <0.113.0> Supervisor vmq_ql_query_sup had child vmq_ql_query started with {vmq_ql_query,start_link,undefined} at <0.12431.18> exit with reason {error,badarg} in context child_terminated

==> log/error.log <==
2017-07-24 21:02:08.371 [error] <0.12499.18> CRASH REPORT Process <0.12499.18> with 0 neighbours exited with reason: no such process or port in call to gen_server:call(<0.1643.0>, {checkin,#Ref<0.0.18874371.205627>,{"MY_HOST",443,hackney_ssl},{sslsocket,{gen_tcp,...},...},...}, infinity) in gen_server:call/3 line 212

==> log/console.log <==
2017-07-24 21:02:08.371 [error] <0.12499.18> CRASH REPORT Process <0.12499.18> with 0 neighbours exited with reason: no such process or port in call to gen_server:call(<0.1643.0>, {checkin,#Ref<0.0.18874371.205627>,{"MY_HOST",443,hackney_ssl},{sslsocket,{gen_tcp,...},...},...}, infinity) in gen_server:call/3 line 212
2017-07-24 21:02:08.372 [info] <0.171.0> alarm_handler: {clear,system_memory_high_watermark}
```
3. I use `vmq_reg:subscribe/4`, `vmq_reg:unsubscribe/4` and publish_function in `vmq_reg:direct_plugin_exports/1` in my plugin

It's unusual happens, sometimes good and sometimes bad, it will be fixed automatic even if I do nothing (also no restarting), and happens again

hope my comment is helpful.

**Comment by larshesel** [comment]
Sorry for the late response here.

We just merged #480 to master which prevents single overloaded/slow queues/sessions from blocking the entire query more than at most 100ms. This is configurable and can be passed to the command like this `vmq-admin session show --client_id=test --is_online=true --rowtimeout=<TimeInMilliseconds>`.

For now only the successful results are shown - if it took too long to get the data from a session or queue it will be logged as a warning. Note, the current log message is not very descriptive but we'll try to improve on that in the next iteration.
